src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go

```diff
@@ -3,106 +3,150 @@ package storage_mining
+	}
+
+	IMPL_TODO() // Determine auxiliary seed input to ensure uniqueness
+	challengedSectorsRandomness := rt.RandomnessWithAuxSeed(
+		filcrypto.DomainSeparationTag_SurprisePoStSampleSectors,
+		rt.CurrEpoch()-node_base.SPC_LOOKBACK_POST,
+		addr.Serialize_Address_Compact(rt.CurrReceiver()),
+	)
+
+	challengedSectors := _surprisePoStSampleChallengedSectors(
+		challengedSectorsRandomness,
+		SectorNumberSetHAMT_Items(st.ProvingSet()),
+	)
+
+	st.Impl().PoStState_ = MinerPoStState_New_Challenged(rt.CurrEpoch(), challengedSectors, numConsecutiveFailures)
 
-	h, st = a.State(rt)
-	// update challenge start epoch
-	st.ChallengeStatus().Impl().OnNewChallenge(rt.CurrEpoch())
 	UpdateRelease(rt, h, st)
-	return rt.SuccessReturn()
+
+	// Request deferred Cron check for SurprisePoSt challenge expiry.
+	provingPeriod := indices.StorageMining_SurprisePoStProvingPeriod()
+	a._rtEnrollCronEvent(rt, rt.CurrEpoch()+provingPeriod, []sector.SectorNumber{})
 }
 
-// called by verifier to update miner state on surprise post submission
-func (a *StorageMinerActorCode_I) ProcessSurprisePoSt(rt Runtime, onChainInfo sector.OnChainPoStVerifyInfo) InvocOutput {
-	TODO() // TODO: validate caller
+// Invoked by miner's worker address to submit a response to a pending SurprisePoSt challenge.
+func (a *StorageMinerActorCode_I) SubmitSurprisePoStResponse(rt Runtime, onChainInfo sector.OnChainPoStVerifyInfo) {
+	h, st := a.State(rt)
+	rt.ValidateImmediateCallerIs(st.Info().Worker())
 
-	if !a._rtVerifySurprisePoSt(rt, onChainInfo) {
-		rt.Abort(
-			exitcode.UserDefinedError(exitcode.PoStVerificationFailed),
-			"PoSt verification failed")
+	if !st.PoStState().Is_Challenged() {
+		rt.AbortStateMsg("Not currently challenged")
 	}
 
-	// Ensure pledge collateral satisfied
-	// otherwise, abort ProcessVerifiedSurprisePoSt
-	rt.Send(addr.StoragePowerActorAddr, ai.Method_StoragePowerActor_EnsurePledgeCollateralSatisfied, []util.Serialization{}, actor.TokenAmount(0))
+	Release(rt, h, st)
 
-	return a._onSuccessfulPoSt(rt)
+	a._rtVerifySurprisePoStOrAbort(rt, onChainInfo)
+	a._rtUpdatePoStState(rt, MinerPoStState_New_OK(rt.CurrEpoch()))
 
+	rt.Send(
+		addr.StoragePowerActorAddr,
+		ai.Method_StoragePowerActor_OnMinerSurprisePoStSuccess,
+		[]util.Serialization{},
+		actor.TokenAmount(0),
+	)
 }
 
-// Commit Sector
-
-// Deals must be posted on chain via sma.PublishStorageDeals before PreCommitSector
-// Optimization: PreCommitSector could contain a list of deals that are not published yet.
-func (a *StorageMinerActorCode_I) PreCommitSector(rt Runtime, info sector.SectorPreCommitInfo) InvocOutput {
-	h, st := a.State(rt)
+//////////////////
+// ElectionPoSt //
+//////////////////
 
-	rt.ValidateImmediateCallerIs(st.Info().Worker())
-	// can be called regardless of Challenged status
+// Called by the VM interpreter once an ElectionPoSt has been verified.
+// Assumes the block reward has already been granted to the storage miner actor.
```

Comment:
We intent to grant block reward at the end of the block, so this will be false. But I can't see anything below that depends on it anyway.

Reply:
Yes, these comments are outdated; I'll remove them in a subsequent PR.

Reply:
#806